fix(@angular-devkit/schematics): filters out non-NPM dependencies (#13059)

dgp1130 opened the pull request:
Non-NPM dependencies in package.json are now filtered out of `ng update`. Dependencies which are linked as tarball or Git references will now be skipped during the update process.

clydin commented on packages/schematics/update/update/index.ts:
```diff
@@ -850,11 +873,16 @@ export default function(options: UpdateSchema): Rule {
+
+    // Filter out non-NPM dependencies by checking if they link to a semver.
+    const npmDeps = new Map(Array.from(allDeps.entries())
+        .filter(([_, source]) => isSimpleSemVer(source)));
```
Consider using [npm-package-arg's](https://github.com/npm/npm-package-arg) `resolve` function to both validate and check for a registry specifier.  This is the logic that npm itself uses.

dgp1130 replied:
Thanks for the tip, all my Google searches were leading me to the `semver` package. Updated to use `npm-package-arg` and also added a test to verify the behavior.

clydin commented on packages/schematics/update/update/index.ts:
```diff
@@ -826,6 +827,13 @@ function _formatVersion(version: string | undefined) {
+// Returns whether or not the given package spec (the value string in a
+// `package.json` dependency) is hosted in the NPM registry.
+function isPkgFromRegistry(name: string, spec: string): boolean {
+  const result = npa.resolve(name, spec);
```
This can throw if the name or spec is invalid.  In those cases, this should probably return false since update will fail when it attempts to analyze the package.

dgp1130 replied:
I was thinking that in the case of an invalid package name, we should just abort and let the update process crash. If this throws, then `package.json` must be malformed or corrupted and would not be usable by NPM either. Is that a case where we would want to the update process to suppress the error and skip that package?

clydin replied:
Barring catastrophic errors such as out of memory exceptions, the CLI itself should ideally never crash.
However, since the update logic is currently wrapped in a schematic, the only way to report an error is to throw an exception so that should work.  The exception messages should probably be augmented and rethrown with additional information/instructions regarding the failure so that they are actionable by the user though.

dgp1130 replied:
"crash" was the wrong word for me to use, but yes I agree. Updated to rethrow a more meaningful error message to the user. This does throw directly from the function rather than emit an error from a returned `Observable`. I think this is ok based on the fact that `_getAllDependencies()` also throws and I'm assuming there is a global error handler which will catch and print the message.

dgp1130 commented:
I had to install `npm-package-arg` which created the `package-lock.json`. Is that not desirable for this project? I figured dependencies just hadn't been added since `package-lock.json` became a thing. Is there a specific reason we're trying to avoid that here?

clydin commented:
`package-lock.json` is npm's lock file.  The repository is setup to use yarn workspaces (https://github.com/angular/angular-cli/blob/master/package.json#L56).  Once the dependency is added to the package.json for one of the repository packages, executing `yarn` within the repository will install in the appropriate locations for the package and update the root lock file.

This package also has a Bazel BUILD file which will need to be updated to include the new dependency as well (https://github.com/angular/angular-cli/blob/master/packages/schematics/update/BUILD#L24).

dgp1130 commented:
Thanks for explaining the lock file, I didn't realize `yarn.lock` applied to every package in the workspace. Removed the extra `package-lock.json` and ran `yarn` to update `yarn.lock`.

Thanks for pointing out the `BUILD` file, I didn't think to look for one. I added what I think is right for my changes, but attempting to build it fails due to missing `@types/npm-package-arg`:

```
douglasparker@gryphon (~/Source/ng-cli)
[19-11-15 10:16:41]$ yarn bazel build //packages/schematics/update
yarn run v1.19.1
$ /usr/local/google/home/douglasparker/Source/ng-cli/node_modules/.bin/bazel build //packages/schematics/update
INFO: Writing tracer profile to '/usr/local/google/home/douglasparker/.cache/bazel/_bazel_douglasparker/fd23369152ca483d39c1138566f81234/command.profile.gz'
INFO: Build option --legacy_external_runfiles has changed, discarding analysis cache.
ERROR: /usr/local/google/home/douglasparker/Source/ng-cli/packages/schematics/update/BUILD:14:1: no such package '@npm//@types/npm-package-arg': BUILD file not found in directory '@types/npm-package-arg' of external reposi
tory @npm and referenced by '//packages/schematics/update:update'
ERROR: Analysis of target '//packages/schematics/update:update' failed; build aborted: no such package '@npm//@types/npm-package-arg': BUILD file not found in directory '@types/npm-package-arg' of external repository @npm
INFO: Elapsed time: 18.172s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (7 packages loaded, 6 targets configured)
    currently loading: @npm//rxjs ... (5 packages)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

My understanding of [Bazel's NodeJS toolchain](https://bazelbuild.github.io/rules_nodejs/install.html) is that it should install these dependencies from the `yarn.lock` file. Is there something else I need to do to include this dependency.

Follow up question, where is this Bazel rule used if not by `yarn build` or `yarn test`? Is this simply an artifact of an ongoing migration to Bazel, or is there some system which uses it instead of `yarn`?

clydin commented on packages/schematics/update/package.json:
```diff
@@ -13,11 +13,13 @@
   "dependencies": {
     "@angular-devkit/core": "0.0.0",
     "@angular-devkit/schematics": "0.0.0",
+    "@types/npm-package-arg": "^6.1.0",
```
These type of dependencies are build time only and should be placed in the root package.json.

This should also fix the bazel build issue

dgp1130 replied:
I see, moved the dependency to the root `package.json`. I also made a dev dependency, since I don't think it would necessary at runtime (though the way `scripts/` are loaded with a require extension, I think they might require types at runtime, but I don't think schematics are loaded in that fashion). `yarn bazel build //packages/schematics/update` now succeeds.

Just out of curiosity, why are build-time dependencies required in the root `package.json`? I would think each package would include all the dependencies necessary to build them. Is this required for Bazel to generate the NPM build targets?

clydin commented on packages/schematics/update/update/index.ts:
```diff
@@ -847,14 +858,23 @@ export default function(options: UpdateSchema): Rule {
+      } catch {
+        // Abort on failure because package.json is malformed.
+        throw new SchematicsException(
+          `Failed to parse dependency "${name}" with spec "${spec}" from`
```
nit: spell out specifier

dgp1130 replied:
Done.

dgp1130 commented:
Merged in to 9.0.x patch branch. Fixed a minor conflict due to a version update to `pacote` which was not present in the patch branch.
